Added Mahalanobis distance to linear regression

sjentsch opened the pull request:
Added Mahalanobis distance (both as a results table and as output data) to linear regression

sjentsch committed:
Automatically created by jmvtools::prepare()

sjentsch committed:
Added implementation of Mahalanobis distance for regression

sjentsch committed:
Automatically created by jmvtools::prepare() and added output to UI definition

sjentsch committed:
Corrected calculation of Mahalanobis output variable

sjentsch committed:
Modified access rights (remove execute for R-files)

sjentsch committed:
Remove non-ASCII characters

raviselker commented on R/linreg.b.R:
```diff
@@ -654,6 +683,13 @@ linRegClass <- R6::R6Class(
+                                   jmvcore::format(.("Mahalanobis distance ({stat}) for the variables in \"Covariates\""), stat = s),
+                                   ''
+                                  )
+            self$results$mahalOV$set(seq(2), titles, descriptions, rep('continuous', 2))
```
At the moment, the Mahalanobis distance does not support adding variables in multiple blocks. If you add two variables in `Block 1`, and one variable in `Block 2` under the Model Builder, only one set of Mahalanobis distance scores is saved to the spreadsheet. However, each model should have its own associated Mahalanobis distance scores (similar to Cooks distance).

sjentsch replied:
Comment (1) on Slack.

sjentsch replied:
Fixed.

raviselker commented on R/linreg.b.R:
```diff
@@ -850,6 +886,37 @@ linRegClass <- R6::R6Class(
+
+            for (i in seq_along(termsAll)) {
+                table <- groups$get(key=i)$dataSummary$mahal
+                if (is.null(mahal[[i]])) {
```
When I only add one variable as a covariate I get the following error:
```
Error in mahal[[i]]: subscript out of bounds

private$.run()
```

sjentsch replied:
I will look into that. I thought I had a if-condition that handles that.

sjentsch replied:
Fixed.

raviselker commented on R/linreg.b.R:
```diff
@@ -850,6 +886,37 @@ linRegClass <- R6::R6Class(
+            for (i in seq_along(termsAll)) {
+                table <- groups$get(key=i)$dataSummary$mahal
+                if (is.null(mahal[[i]])) {
+                    table$setNote("row", .("Mahalanobis distance can only be calculated for models with two or more dependent variables."))
```
It would be nicer if the user gets a `Notice` (see utils `setAnalysisNotice`) function when the Mahalanobis distance cannot be calculated.

sjentsch replied:
I agree and will change that. What you propose makes more sense.

sjentsch replied:
I tried, however, I don't think it improves things. Since the Mahalanobis distance can only be calculated with two or more variables, one would see this error message every time one runs a hierarchical regression where the first block contains only one variable. This might be annoying for the users?

sjentsch replied:
It would make sense if I could attach it to the group / the block. If this is possible, let me know, I could not figure out how.

raviselker replied:
Yeah, that's indeed how I would do it. You can manually add a notice by doing the following (I tested it, and it works):
```R
                    notice <- jmvcore::Notice$new(
                        options = self$options,
                        name = 'warningMessage',
                        type = jmvcore::NoticeType$STRONG_WARNING
                    )
                    notice$setContent(.("Mahalanobis distance can only be calculated for models with two or more independent variables."))
                    groups$get(key=i)$insert(1, notice)
```

sjentsch replied:
Using a notice is implemented now (see above).

raviselker commented on R/linreg.b.R:
```diff
@@ -850,6 +886,37 @@ linRegClass <- R6::R6Class(
+                        "row",
+                        jmvcore::format(.("Filter syntax: {filter}"),
+                            filter = paste(paste0("ROW() != ", mahal[[i]][selRow, "row"]), collapse = " and "))
+                    )
```
I'm not a fan of this note because:
1. I think it would be easier (and more robust) to use the saved Mahalanobis distance to filter out rows
2. This note only makes sense if you use jamovi, and not if you're using `jmv` through R directly

jonathon-love replied:
> I think it would be easier (and more robust) to use the saved Mahalanobis distance to filter out rows

i think this is necessary because you can't filter on output variables ... it creates a circular dependency ... you apply a filter which changes the analysis, which changes the output column, which changes the filter, which changes the analysis ... and so on.

sjentsch replied:
Point 2, I can't really solve (but people using `jmv` on R may be knowledgable enough to “translate” jamovi-filter into something that can be used in R.
The comment of jonathon on point 1 is exactly my point (you can't use the output variables for filtering).

sjentsch replied:
But I am certainly will listen if there is a better way of solving this.

jonathon-love replied:
longer term, we're going to support multiple "sheets" within a "project", and one would be able to make a second "sheet" containing the data filtered by m9s so this filter "formula output" won't be necessary then, but we'll have to keep it in because people don't cope when you remove things. ... and of course if we keep it in, people will assume that's the "recommended way" to do it.

how about we approach this a different way @sjentsch, what if you instruct people to copy the m9s output variable into a new data variable, and then filter on that?

i appreciate it's hackish and ugly ... but it might be the lesser of several evils.

thoughts?

raviselker replied:
Yeah, I'd be fine with this workaround

sjentsch replied:
I removed the filter syntax. This will enable people to either copy the output columns and use that or to just create a filter syntax by writing the line numbers with the outliers manually (like I did).
@raviselker: I asume the line numbers match those of the dataset when I use `self$data`?

jonathon-love replied:
if you've got filtered rows, then these won't appear in `self$data` ... so if the first row was filtered out of the spreadsheet, the second row in the spreadsheet would appear in the first row in `self$data`, the third row in the spreadsheet would appear in the second row of `self$data`, etc.

you can get the actual row numbers with `row.names(self$data)`

sjentsch replied:
Yes, but I actually do this. What is in the column `rows`  gets assigned `return(data.frame(row = row.names(data), chisq = mahalChiSq, p = pchisq(mahalChiSq, ndep, lower.tail = FALSE)))` (l. 302) with the data coming from `.cleanData()`  which assigns the correct `row.names` (l. 1456). It's perhaps not obvious because of the three steps, but in large parts similar to what Ravi (or whoever wrote these parts of `linreg.b.R`) does.

raviselker replied:
So the `private$.getDataRowNums()` gives the row numbers of the `self$dataProcessed` data set. However, for some reason you use `private$.cleanData(naSkip=jmvcore::toB64(self$options$dep))` instead. Therefore, it's possible that the length of the `getDataRowNums` and the cleaned data do not match, which would result (I assume) in an error.

So a solution would be etiher one of these:
1. use `self$dataProcessed` instead (just like I've done in the cooks distance)
2. get the rownames from the mahal distance itself

Unless there's a good reason to not use `self$dataProcessed`, I'd go for option `1`

sjentsch replied:
Oh, I did not know. I just used it because you use it a couple of rows up in `.computePredicted`.

raviselker commented on R/linreg.b.R:
```diff
@@ -285,6 +293,27 @@ linRegClass <- R6::R6Class(
+            for (i in seq_along(self$models))
+                mahal[[i]] <- mahal2List(data, attr(self$models[[i]]$terms, "term.labels"))
+            
+            mahal[[i + 1]] <- mahal2List(data, jmvcore::toB64(self$options$dep))
```
Why do you do this? Wouldn't this always retun `NULL`?

sjentsch replied:
This may go away if we create the output variables model-wise. It is currently there to provide the output variables for all “Covariates” in the model.

sjentsch replied:
Resolved because I now calculate Mahalanobis distances for each block.

raviselker commented on R/linreg.b.R:
```diff
@@ -850,6 +886,37 @@ linRegClass <- R6::R6Class(
                 table$setRow(rowNo=1, values=cooks[[i]])
             }
         },
+        .populateMahalTable = function() {
```
This table now works differently from the cooks table, which is more of a summary of the distances. The nice thing about a summary table is that you get a quick overview of the distribution of values, and if you want to do something based on these values it makes more sense to use the saved mahalanobis distances. So I'd prefer a similar table to the Cook's distance would be given here.

raviselker commented on R/linreg.b.R:
```diff
@@ -1020,6 +1087,13 @@ linRegClass <- R6::R6Class(
                     self$results$cooksOV$setValues(index=i, self$cooks[[i]])
             }
+
+            if (self$options$mahalOV && self$results$mahalOV$isNotFilled()) {
```
This only works for a single model, not when you use the model builder and add multiple blocks

sjentsch replied:
Fixed.

raviselker commented on tests/testthat/testlinreg.R:
Could you exclude file mode changes from the git commit (i.e., by `git config core.fileMode false`)

I think each added feature should test it. I think the tests should also represent different use cases, such as multi block analyses, and fewer than 2 dependent variables to make sure these use cases are done correctly

sjentsch replied:
Yes, and I should perhaps not have changed them. I just thought these are R-files and the should not be executable. Just drop the commit, if you disagree.

raviselker commented on tests/testthat/testmancova.R:
Could you exclude file mode changes from the git commit (i.e., by git config core.fileMode false)

jonathon-love replied:
these changes are in a separate commit, so that seems OK to me?

raviselker replied:
Yeah, but this feature shouldn't have to touch the `testmancova` file, so it's a bit strange that it's changed in this MR. For the other file (the `testlinreg` I'm fine with keeping it)

raviselker reviewed (changes requested):
Hi @sjentsch, I think adding the mahalanobis distance measures would be a good idea. However, I think you'd need to address a couple of issues for it to be merged. If you need some help, lmk through slack

sjentsch committed:
Implement output for Mahalanobis per model / block

sjentsch committed:
Change Mahalanobis distance summary to showing (one-line) summary statistics and cases to exclude

sjentsch committed:
Fixing errors in calculating / showing Mahalanbis distance in hierarchical regression

sjentsch committed:
Bug fix: Do not include factors in the calculation of the Mahalanobis distance

sjentsch committed:
Added unit tests for Mahalanobis distances

raviselker commented on R/linreg.b.R:
```diff
@@ -285,6 +293,21 @@ linRegClass <- R6::R6Class(
         },
+        .computeMahal = function() {
+            mahal <- vector(mode = "list", length(self$nModels))
+            data <- private$.cleanData(naSkip=jmvcore::toB64(self$options$dep))
```
Is there a reason you don't just use the `dataProcessed` active binding?

raviselker commented on R/linreg.b.R:
```diff
@@ -285,6 +293,21 @@ linRegClass <- R6::R6Class(
+            data <- private$.cleanData(naSkip=jmvcore::toB64(self$options$dep))
+            for (i in seq_along(self$models)) {
+                cov <- attr(self$models[[i]]$terms, "term.labels")
+                cov <- intersect(cov, jmvcore::toB64(self$options$covs))
```
Are factors not relevant for calculating the mahalanobis distance? At this moment it's calculated only based on the covariates.

sjentsch replied:
No, it is only for continuous variables.

sjentsch committed:
Changed warning about min. number of variables into notice and removed filter syntax

sjentsch committed:
Finished implementing Mahalanobis distance (incl. notice)

jonathon-love committed:
Updated i18n

raviselker committed:
linReg: improve singularity error message

With this commit, an warning notice is added, if one of the models
in the analysis has a singularity issue. Also fixes an indexing problem
that appears when a model contains aliased coefficients.

sjentsch committed:
Changed private$.cleanData into self

raviselker reviewed (approved):
I checked the changes, and they address the issues I originally had so this can me merged
